lib/sparkql/parser_compatibility.rb

```diff
@@ -205,6 +205,10 @@ def check_type!(expression, expected, supports_nulls = true)
       expression[:type] = :datetime
       expression[:cast] = :date
       return true
+    elsif expected == :date && expression[:type] == :datetime && datestr = expression[:value].match(/^(\d{4}-\d{2}-\d{2})/)
```

Comment:
Validated at least SOME assurance that we're dealing with a date. Nulls are handled up above, so I think we're in the clear.